File: C++/156_Binary Tree Upside Down.cpp
class Solution {
public:
    TreeNode *upsideDownBinaryTree(TreeNode *root) {
        if (root == NULL || root->left == NULL)
            return root;
        flip(root);
        return res;
    }
private:
    TreeNode *res;
    TreeNode *flip(TreeNode *root) {
        if (root->left == NULL) {
            res = root;
            return root;
        }
        TreeNode *p = flip(root->left);
        p->right = root;
        p->left = root->right;
        root->left = root->right = NULL;
        return root;
    }
};

File: Python/156_Binary Tree Upside Down.py
class Solution:
    # @param {TreeNode} root
    # @return {TreeNode}
    res = None
    def upsideDownBinaryTree(self, root):
        if root is None:
            return root
        self._upsideDown(root, None)
        return self.res

    def _upsideDown(self, root, parent):
        if root is None:
            return
        self._upsideDown(root.left, root)
        if self.res is None:
            self.res = root
        if parent is not None:
            root.left = parent.right
            root.right = parent
        else:
            root.left = root.right = None